chore(api): upgrade to api-platform 3.0

usu opened the pull request:
Superseded #2984 (had to create a new PR, because test workflows didn't run on the old one anymore)

### ToDo
- [x] check performance timings (sanity check before/after)
- [x] refactor/fix tests for state processors
- [x] fix phpstan/psalm warning
- [x] review deprecation warnings
- [x] check if in frontend everything is still working
- [x] IRIs of contentNodes not yet consistent (needs the following api-platform fix: https://github.com/api-platform/core/pull/5261)

### Stuff for separate PRs
- Review changes done in https://github.com/api-platform/api-platform and take over what makes sense
- Review if we want to support multiple ApiResources in our various extensions (UriTemplateFactory, etc.)
- ~~Register upstream issues~~
- Plural naming for /reset_password endpoint
- Test coverage for data providers

usu committed:
convert legacy interfaces/factories

usu committed:
run  + formatting + clenaup + revert unnecessary changes

usu committed:
first working test: testGetSingleDayIsAllowedForMember

usu committed:
upgrade to api-platform 3 + fix pointers to old ApiPlatcorm\Core namespace + fix the same test: testGetSingleDayIsAllowedForMember

usu committed:
exclude DataPersister/DataProvider from service discovery (unfixed code)

usu committed:
fix oauth

usu committed:
fix getExamplePayload for testing

usu committed:
merge devel

usu committed:
upgrade to api-platform 3.0.2

usu committed:
fix TranslationNormalizer and Skolem IRI BC change

usu committed:
exampe StateProcessor for Activity

usu committed:
example state processors for CampCollaboration

usu committed:
abstraction of property changes via AbstractPersistProcessor

usu committed:
merge devel

usu committed:
upgrade to api-platform/core 3.0.3

usu committed:
fix CreateCampCollaborationTest

usu committed:
bind  to abstract class

usu committed:
state processors for Camp and Period

usu committed:
state processors for Category

usu committed:
state processors for Profile

usu committed:
state processors for User

usu committed:
state processors for Invitation DTO

usu committed:
state processors for ResetPassword

usu committed:
state processors for content nodes

usu committed:
remove dead code

usu committed:
fix api tests for MaterialNode, MultiSelect and CampCollaboration

usu committed:
fix api tests for Invitation DTO

usu committed:
fix MaterialItemDenormalizer

usu committed:
upgrade api-platform/core to 3.0.4

usu committed:
remove Get operation from the general ContentNode endpoint

usu committed:
fix ValidateDeleteRequestsListener

usu committed:
fix tests for ExpressionDateTimeFilter

usu committed:
downgrade api-platform to 3.0.3

usu committed:
fix tests for UriTemplateFactory

usu committed:
fix tests for RelatedCollectionLinkNormalizer

usu committed:
fix tests for UriTemplateNormalizer

usu committed:
fix PreventAutomaticEmbeddingPropertyMetadataFactoryTest

usu committed:
fix MailServiceTest

usu committed:
fix DTO uri

usu committed:
merge devel

usu committed:
fix reset_password endpoint

usu committed:
fix various psalm/phpstan/cs-fixer errors

usu committed:
consistent state processor handling for DTOs

usu commented:
@ecampcore I think this PR is ready to be looked at. The only larger pending point is to transform the tests from data persisters to state processors. All other unit tests and API tests are running fine. There are some discussion points (marked in the code), so probably better to already start looking into them. The changes to API tests are minor, but there are some. Nothing of these should break the frontend, but please still have a look at them to verify if you agree.

BacLuc commented on api/tests/bootstrap.php:
```diff
@@ -4,6 +4,8 @@
 
 require dirname(__DIR__).'/vendor/autoload.php';
 
+DG\BypassFinals::enable(); // allows mocking of final classes
```
why do we need this?

usu replied:
`ApiPlatform\Doctrine\Orm\Filter\SearchFilter` is now newly a final class. In `RelatedCollectionLinkNormalizerTest` we create a test mock for a SearchFilter. However, creating a mock for a final class throws an error.

This is one of various solutions i found on:
https://tomasvotruba.com/blog/2019/03/28/how-to-mock-final-classes-in-phpunit/

carlobeltrame replied:
If that is the only reason, I have adjusted the test to not rely on mocking `SearchFilter` anymore: https://github.com/usu/ecamp3/pull/13
I think it's dangerous to allow to override final classes, and would like to avoid it if possible, even if it's a little more inconvenient to write tests this way.

carlobeltrame commented on api/composer.json:
```diff
@@ -83,12 +84,20 @@
             "App\\": "src/"
-        }
+        },
+        "exclude-from-classmap": [
+            "**/DataPersister/**",
+            "**/DataProvider/**"
+        ]
```
What does this do?

usu replied:
This is the legacy business logic. Will remove.

usu replied:
This is now removed

carlobeltrame commented on api/config/services.yaml:
```diff
@@ -20,13 +31,12 @@ services:
             - '../src/Entity/'
             - '../src/Kernel.php'
             - '../src/Tests/'
+            - '../src/DataPersister/'
+            - '../src/DataProvider/'
```
Why exclude these? Aren't they needed as services? Is it because we explicitly name them in the Entities now? Is there a downside to having these classes as services? I thought Symfony was smart enough to optimize the service container based on actually used services.

usu replied:
This folders are now empty (legacy persisters and providers). All new code is in the folder "State". Will remove.

usu replied:
This is removed now

carlobeltrame committed:
Use more mocking instead of allowing to override final classes

If a package vendor specifies a class as final, they really do not
expect the class to ever be overridden. Also, adding this package
to require-dev allows us to do so in a dev environment, even
accidentally outside of tests, and that code will only fail when
running in production.
Instead of trying to enable mocking SearchFilter, we can just
instantiate it and mock a little more in another place.
The test testNormalizeDoesntReplaceWhenStrategyIsNotExact was flawed,
because the search strategy is always forced to be "exact" when
filtering by an association. For this reason, I removed that test,
because it does not make sense to test a primitive field filter in the
context of RelatedCollectionLinkNormalizer. This normalizer is only
concerned with associations, not with primitive fields.

carlobeltrame commented on api/src/DTO/Invitation.php:
```diff
@@ -12,51 +18,39 @@
+    operations: [
+        new Get(
+            provider: InvitationProvider::class,
+            uriTemplate: '/invitations/{inviteKey}/find{._format}', // TO DISCUSS: Wouldn't '/{inviteKey}{._format}' be more REST-like
```
Sure, I agree. And we could use PATCH for accept, and DELETE for reject. Using DELETE will need changes in the frontend as well though.

usu replied:
Let's discuss in the next meeting. Technically, both the accept and reject modify the underlying campCollaboration but delete the invitation (delete in the sense that the invitation is not accessible anymore afterwards).

BacLuc replied:
i would not change that in this PR, but save it for a later PR

manuelmeister replied:
## Core Meeting Decision

- [ ] remove find
- [ ] at a later point move to delete instead of patch

usu replied:
I kept the `/find` for the moment, it didn't intervene too much. We can clean up in a separate PR. I will create a master issue with all the remaining points after merging this.

carlobeltrame commented on api/src/DTO/ResetPassword.php:
```diff
@@ -2,38 +2,43 @@
+    operations: [
+        new Get(
+            provider: ResetPasswordProvider::class,
+            uriTemplate: '/reset_password/{id}{._format}', // TO DISCUSS: default uri would be /reset_password (plural). Shall we keep or fix?
```
Do you mean `reset_passwords` with an `s` in the end? Hm not sure... Would be consistent I guess, and should be trivial to change.

usu replied:
Ah sorry, yes, `reset_passwords` with a plural `s`

BacLuc replied:
it was reset_password before. I would change that in a later PR

carlobeltrame commented on api/src/Entity/Activity.php:
```diff
@@ -20,34 +27,36 @@
+            normalizationContext: ['groups' => ['read', 'Activity:ActivityResponsibles', 'Activity:ScheduleEntries']],
+            security: 'is_authenticated()'
+        ),
+        new Post(
```
How did you order the operations? Any consistent pattern? I could imagine a few ways, e.g. alphabetic, in the order of CRUD, readonly before write operations, collection operations before item operations (like Swagger does it), but none seem to be applied.

usu replied:
I didn't, rector did. Thas was the only part that went fully automatic 😅 

I'm not 100% sure, but looking at the entities (for example CampCollaboration) I think rector preserved the order of operations but first listed all `itemOperations` and then `collectionOperations`.

carlobeltrame commented on api/src/Entity/Activity.php:
```diff
@@ -75,7 +84,7 @@ class Activity extends BaseEntity implements BelongsToCampInterface {
     #[ApiProperty(
         writableLink: true,
-        example: '[{ "period": "/periods/1a2b3c4a", "endOffset": 1100, "startOffset": 1000 }]',
+        example: '[{ "period": "/periods/1a2b3c4a",  "start": "2023-05-01T15:00:00+00:00", "end" => "2023-05-01T16:00:00+00:00" }]',
```
```suggestion
        example: [[ "period" => "/periods/1a2b3c4a",  "start" => "2023-05-01T15:00:00+00:00", "end" => "2023-05-01T16:00:00+00:00" ]],
```
Nice catch, but you are mixing `:` and `=>`, and we can write this as PHP objects to get a more accurate example value (array of objects rather than JSON string)

usu replied:
Fixed in https://github.com/ecamp/ecamp3/pull/3176/commits/dfb0b16cef2304a2b95128b69eebea07f3699e82

carlobeltrame commented on api/src/Entity/ActivityResponsible.php:
```diff
@@ -17,22 +21,27 @@
     message: 'This campCollaboration (user) is already responsible for this activity.',
 )]
-#[ApiFilter(SearchFilter::class, properties: ['activity', 'activity.camp'])]
+#[ApiFilter(filterClass: SearchFilter::class, properties: ['activity', 'activity.camp'])]
```
Is the `filterClass:` param name required now?

usu replied:
Was added automatically by rector, didn't bother to check if this is required. If it's working without, then I'm fine with both variants.

carlobeltrame commented on api/src/Entity/ContentNode/MultiSelect.php:
```diff
@@ -2,38 +2,49 @@
-            'security' => 'is_granted("CAMP_MEMBER", object) or is_granted("CAMP_MANAGER", object)',
-            'validation_groups' => ['Default', 'update'],
-        ],
-        'delete' => ['security' => '(is_granted("CAMP_MEMBER", object) or is_granted("CAMP_MANAGER", object)) and object.parent !== null'], // disallow delete when contentNode is a root node
```
Is there a way to keep this comment? Or do we at least have a clearly named test which specifically tests this?

usu replied:
Fixed in https://github.com/ecamp/ecamp3/pull/3176/commits/dfb0b16cef2304a2b95128b69eebea07f3699e82

carlobeltrame commented on api/src/Entity/ScheduleEntry.php:
```diff
@@ -106,7 +111,7 @@ class ScheduleEntry extends BaseEntity implements BelongsToCampInterface {
      * API currently does not enforce this.
      */
-    #[ApiProperty(default: 0, example: 0.6)]
+    #[ApiProperty(default: '0', example: 0.6)]
```
Is that change really correct? Do we have a test which submits a string in this field?

usu replied:
The default property is currently defined as `@param string|null $default`, so phpstan/psalm will complain.

We're not using this very often in our code base. Do you know what this is exactly used for? Couldn't find any documentation on the consequence of specifying the default property.

carlobeltrame replied:
Hm not sure actually. During the review, I confused it with `example`, which is still a float. Example is definitely displayed in swagger, not sure about the default. Is it maybe used in Create operations when the user sends no value for left? Maybe we need a test for that then?
On the other hand, with all three picasso displays using vuetify's stacking algorithm, we could also just create an issue to remove this field entirely, and call it a day.

usu replied:
There's actually already an issue for that: https://github.com/ecamp/ecamp3/issues/2368

We're currently using the default property at 1 other location:
https://github.com/ecamp/ecamp3/blob/devel/api/src/Entity/Category.php#L146
https://github.com/ecamp/ecamp3/blob/devel/api/src/Entity/CampCollaboration.php#L185

So maybe we want to remove it from there, if we're not really sure what effect this property has.

manuelmeister replied:
## Core Meeting Decision
Remove default as we set default values in php itself.

usu replied:
removed in https://github.com/ecamp/ecamp3/pull/3176/commits/47a3838f2a7a3870ac92c580e8117bcc9ac38c0b

carlobeltrame commented on api/src/Metadata/Resource/Factory/GraphQLSecurityResourceMetadataFactory.php:
```diff
@@ -1,46 +0,0 @@
-use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
-use ApiPlatform\Core\Metadata\Resource\ResourceMetadata;
-
-class GraphQLSecurityResourceMetadataFactory implements ResourceMetadataFactoryInterface {
```
Did we agree to completely abandon GraphQL? In #2559 I read "disable graphql only on prod", and also in https://github.com/ecamp/ecamp3/commit/8cc39ac0e4d5c52a5aa39cc44fb8fe56901953a6 I only implemented it for "production" environments, i.e. deployments.
For using GraphQL, this class is still required even after api platform 2.7, unless you want to separately write all security rules for GraphQL again.
Also don't forget to restore the test in case you are restoring this class.

usu replied:
I didn't really remember we had this still enabled for dev, so I thought it's just some legacy code and I saved a bit of time by not migrating this.

I can migrate and bring it back. Honestly, not sure about the value, though. I'm not sure, this code will make it to production exactly like this.

manuelmeister replied:
## Core Meeting Decision

Move into own issue, so we don't lose all knowledge how GraphQL Security works

carlobeltrame commented on api/src/Metadata/Resource/Factory/UriTemplateFactory.php:
```diff
@@ -2,27 +2,37 @@
 use Psr\Container\ContainerInterface;
 
+/**
+ * Creates templates URIS for resources.
```
```suggestion
 * Creates templated URIs for resources.
```

But probably could be worded more clearly:
```suggestion
 * Given an entity class, creates an URI template (link with placeholders) for accessing this type of entity in the API.
```

usu replied:
Fixed in https://github.com/ecamp/ecamp3/pull/3176/commits/dfb0b16cef2304a2b95128b69eebea07f3699e82

carlobeltrame commented on api/src/Metadata/Resource/OperationHelper.php:
```diff
@@ -0,0 +1,35 @@
+use ApiPlatform\Metadata\Resource\ResourceMetadataCollection;
+
+class OperationHelper {
+    public static function findOneByType(ResourceMetadataCollection $collection, string $className): ?Operation {
```
Cool. Unit test? Extracting code implies that it could be reused soon in different places, which means we should make sure that this method fulfills a contract.

usu replied:
Yes, will add

usu replied:
Implemented in https://github.com/ecamp/ecamp3/pull/3176/commits/548576161911ce9afbaa781df027389a328078b7

carlobeltrame commented on api/src/Serializer/Normalizer/RelatedCollectionLinkNormalizer.php:
```diff
@@ -170,7 +169,14 @@ public function getRelatedCollectionHref($object, $rel, array $context = []): st
+        $resourceMetadataCollection = $this->resourceMetadataCollectionFactory->create($relatedResourceClass);
+        $operation = OperationHelper::findOneByType($resourceMetadataCollection, GetCollection::class);
+
+        if (!$operation) {
+            throw new UnsupportedRelationException('The resource '.$relatedResourceClass.' does not implement GetCollection() operation.');
+        }
```
Nice. Could you add a unit test for this condition? I have so far aimed for 100% test coverage on all early exit conditions in these complicated normalizers.

usu replied:
Implemented in https://github.com/ecamp/ecamp3/pull/3176/commits/382ac5d2ad758ddb992e6b4fa0ce215dbead1560

carlobeltrame commented on api/src/Serializer/Normalizer/UriTemplateNormalizer.php:
```diff
@@ -52,7 +52,7 @@ public function normalize($object, $format = null, array $context = []): array|s
         $result['_links']['oauthPbsmidata'] = ['href' => $this->urlGenerator->generate('connect_pbsmidata_start').'{?callback}', 'templated' => true];
         $result['_links']['oauthCevidb'] = ['href' => $this->urlGenerator->generate('connect_cevidb_start').'{?callback}', 'templated' => true];
-        $result['_links']['resetPassword'] = ['href' => $this->urlGenerator->generate('api_reset_passwords_post_collection').'{/id}', 'templated' => true];
+        $result['_links']['resetPassword'] = ['href' => $this->urlGenerator->generate('_api_/reset_password{._format}_post').'{/id}', 'templated' => true];
```
What is this `_api_`? Why do you specify a URL here? Is the route name no longer available?

usu replied:
It looks like a URL, but it actually is the operation name 😄 

This is new naming scheme from api-platform => _api_ + uriTemplate + method
(and after [this issue](https://github.com/api-platform/core/issues/5203) is resolved also routePrefix)

carlobeltrame replied:
I see, and why didn't you have to change `connect_cevidb_start` etc? Is it because we give explicit names to these?

usu replied:
Yes, correct. `connect_cevidb_start` is coming from a custom controller and the route name is directly specified via attribute:
https://github.com/ecamp/ecamp3/blob/devel/api/src/Controller/CevidbController.php#L14

For API resources, api platform will generate a route name for each operation (which also serves as the operation name). This could be overwritten with the `name` property of any `Operation` attribute, but I guess we're ok with the automatic generated ones.

carlobeltrame commented on api/src/Serializer/PreventAutomaticEmbeddingPropertyMetadataFactory.php:
```diff
@@ -58,26 +58,35 @@ public function __construct(private PropertyMetadataFactoryInterface $decorated)
 
-        return new PropertyMetadata(
-            $propertyMetadata->getType(),
+        return new ApiProperty(
```
`$this->decorated` no longer returns a PropertyMetadata, so you probably should rename the local variable to `$apiProperty`.

Also, `ApiProperty` implements a builder pattern: https://github.com/api-platform/core/blob/main/src/Metadata/ApiProperty.php
So you can reuse the `$apiProperty` from `$this->decorated->create(...)` and `return $apiProperty->withReadableLink(null)->withWritableLink(null);`
Should be way clearer this way what we actually do in this class.

usu replied:
Using the builder doesn't work. `withReadableLink` and `withWritableLink` don't accept `null` as argument.

Renaming is done in https://github.com/ecamp/ecamp3/pull/3176/commits/5e83943cfc2001c520ea438cfea130774ac2831e

carlobeltrame replied:
-.- thanks

carlobeltrame commented on api/src/Service/MailService.php:
```diff
@@ -19,32 +20,42 @@ class MailService {
-        } catch (TransportExceptionInterface $e) {
-            throw new \RuntimeException($e);
+    public function sendInviteToCampMail(CampCollaboration $campCollaboration): void {
+        if (CampCollaboration::STATUS_INVITED == $campCollaboration->status && $campCollaboration->getEmail()) {
```
I think this check and the calculation of the `$byUser`, `$camp` etc. belongs in the StateProcessor instead. If the check does not succeed here, we just silently don't send any email out without any error in the API response. The MailService should just be responsible for sending emails (based on a handful of templates), not for dealing with camp collaborations and statuses in my opinion.

If you insist on keeping this logic in this class, please write tests for it.

BacLuc replied:
+1

usu replied:
reverted to previous setup with https://github.com/ecamp/ecamp3/pull/3176/commits/546cc634639bc05125b7dfaae05b29a8f12ccc06

carlobeltrame commented on api/src/State/ContentNode/StoryboardPersistProcessor.php:
```diff
@@ -1,35 +1,31 @@
+        $data = parent::onBefore($data, $operation, $uriVariables, $context);
 
-        if (null === $data->data) {
+        if ($operation instanceof Post && null === $data->data) {
```
Wouldn't it be more consistent to create a separate CreateProcessor and UpdateProcessor here to distinguish between these two operations? You have done the same for CampCollaboration, ResetPassword and User

usu replied:
When the business logic was the same for update & create or very similar, I tried to avoid creating separate processors but keep it one PersistProcessor. In this specific case, this made it  easier to reuse `sanitizeData`.

But yeah, we can also make 2 processors and then outsource `sanitizeData` into a trait or a separate class.

manuelmeister replied:
## Core Meeting Decision

Namespacing: Provider and Processer are left inside state as is

carlobeltrame commented on api/src/State/InvitationRejectProcessor.php:
```diff
@@ -0,0 +1,36 @@
+        $campCollaboration->inviteKey = null;
+        $campCollaboration->inviteKeyHash = null;
+
+        $this->em->flush();
```
This line is new, really needed?

usu replied:
The way it's designed right now ist that the DTO processors (like `InvitationRejectProcessor`) don't extend from `App\State\Util\AbstractPersistProcessor` (which in turn decorates the default [`ApiPlatform\Doctrine\Common\State\PersistProcessor`](https://github.com/api-platform/core/blob/3.0/src/Doctrine/Common/State/PersistProcessor.php)).

This prevents the special hack that is needed, where the `onBefore` needs to return a `CampCollaboration` (or any other valid Doctrine Entity) and then onAfter or an overridden process method needs to ensure an `Invitation` object is returned and not a `CampCollaboration` entity.

As a consequence, because this processor doesn't decorate the default ORM PersistProcessor, the persisting has to be done manually (e.g. `flush()`). To me, this seems more easy to read/understand as it's more explicit of what's happening. But happy to discuss.

carlobeltrame commented on api/src/State/InvitationAcceptProcessor.php:
```diff
@@ -0,0 +1,40 @@
+        $campCollaboration->inviteKeyHash = null;
+        $campCollaboration->inviteEmail = null;
+
+        $this->em->flush();
```
This line is new, really needed?

carlobeltrame commented on api/src/State/InvitationProvider.php:
```diff
@@ -1,20 +1,17 @@
-class InvitationDataProvider implements ItemDataProviderInterface, RestrictedDataProviderInterface {
-    use SerializerAwareDataProviderTrait;
-
+class InvitationProvider implements ProviderInterface {
```
Is it normal that the StateProcessors and StateProviders are now in the same namespace?

usu replied:
This is how it's indicated in the examples from the documentation under https://api-platform.com/docs/core/state-providers/ and https://api-platform.com/docs/core/state-processors/. And this also the folder that is used by the MakerBundle when using `bin/console make:state-processor`.

But I guess we are free to change if it makes sense. One option would be to segregate processors and providers. Another option would be to segregate by endpoint to have all processors and providers of a given resource in a separate folder/namespace.

carlobeltrame commented on api/src/State/InvitationAcceptProcessor.php:
```diff
@@ -0,0 +1,40 @@
+use Symfony\Component\PasswordHasher\Hasher\PasswordHasherFactoryInterface;
+use Symfony\Component\Security\Core\Security;
+
+class InvitationAcceptProcessor implements ProcessorInterface {
```
Does this endpoint still return the Invitation? Before we had to take special care not to return the CampCollaboration.

usu replied:
Yes, the process method in the InvitationAcceptProcessor returns an Invitation object (also ensured in the function signature).

```php
    /**
     * @param Invitation $data
     */
    public function process(mixed $data, Operation $operation, array $uriVariables = [], array $context = []): Invitation {
```

carlobeltrame commented on api/src/State/PeriodPersistProcessor.php:
```diff
@@ -1,47 +1,32 @@
-    public function beforeCreate($data): Period {
-        static::updateDaysAndScheduleEntries($data);
+    public function onBefore($data, Operation $operation, array $uriVariables = [], array $context = []): Period {
+        static::updateDaysAndScheduleEntries($data, $context['previous_data']);
```
Does this really work even in the Create case? Do I remember correctly that API platform fills this `$context['previous_data']` for us?

usu replied:
Yes, api-platform populates `$context['previous_data']`. This was already the case in 2.6, so I'm not 100% sure why we haven't used it before.

In create case, `$context['previous_data']` is `null` and tests run fine. So yes, it should work, but anyway let's double check the test cases are properly set up.

carlobeltrame commented on api/src/State/Util/AbstractPersistProcessor.php:
```diff
@@ -0,0 +1,80 @@
+    /**
+     * For side effects after processing the object.
+     */
+    public function onAfter($data, Operation $operation, array $uriVariables = [], array $context = []): void {
```
Can this no longer process the data? I thought before we could also manipulate the returned data after it was stored in the DB..?

usu replied:
Nope, currently not possible in the current codebase:
https://github.com/ecamp/ecamp3/blob/devel/api/src/DataPersister/Util/AbstractDataPersister.php#L82

carlobeltrame commented on api/tests/Api/ContentNodes/ContentNode/ReadContentNodeTest.php:
```diff
@@ -12,60 +12,60 @@
-        $contentNode = static::$fixtures['columnLayoutChild1'];
-        $children = new ArrayCollection($contentNode->getChildren());
-        $childrenIris = $children->map(fn (ContentNode $contentNode) => $this->getIriFor($contentNode))->toArray();
+    // public function testGetSingleContentNodeIsAllowedForCollaborator() {
+    //     /** @var ContentNode $contentNode */
+    //     $contentNode = static::$fixtures['columnLayoutChild1'];
+    //     $children = new ArrayCollection($contentNode->getChildren());
+    //     $childrenIris = $children->map(fn (ContentNode $contentNode) => $this->getIriFor($contentNode))->toArray();
```
Is this endpoint no longer supported? If so, please remove the tests, and make sure that the frontend never uses it. Or maybe get this endpoint to return a useful and understandable "not implemented" error for developers which try to use it?

usu replied:
I kept the code until we take a decision whether to keep it or not. See comment in `ContentNode.php`:
https://github.com/ecamp/ecamp3/pull/3176/files#diff-93cc61a23d08cab7232453b3406546f4512315d55c95ecb04aa458e10a7395acR38

manuelmeister replied:
## Core Meeting Decision

- remove get operation

usu replied:
Removed with https://github.com/ecamp/ecamp3/pull/3176/commits/e489ad8e4ed0f1e2446264dcc9cad5695824fb8f

carlobeltrame commented on api/tests/Api/ContentNodes/ColumnLayout/CreateColumnLayoutTest.php:
```diff
@@ -12,7 +12,7 @@ class CreateColumnLayoutTest extends CreateContentNodeTestCase {
         parent::setUp();
 
-        $this->endpoint = '/content_node/column_layouts';
+        $this->endpoint = '/column_layouts';
```
Why did you change this? What was wrong with the old way? In tests, whenever possible, I like working with explicit, readable, constant literal values which make it easy to understand what the test does. This way the tests can be a kind of living documentation (one which we are actually forced to keep up to date).

usu replied:
I needed to separate `routePrefix` (=`"/content_node"`) and `uriTemplate` (=`"/column_layouts"`) in order to compile the correct operationName in `getExamplePayload`:
https://github.com/ecamp/ecamp3/pull/3176/files#diff-1c80e80e582855ae7b4bbbdfce8894f47dcb3ef2a8e5093dfef03cba5cebf9c7R148

This was needed, because currently in api-platform `routePrefix` is not part of the operation name. I earlier thought this was by purpose, but later realized this is a bug. It's registered upstream and a PR already exists. After this is merged/released, I can revert the change.

https://github.com/api-platform/core/issues/5203
https://github.com/api-platform/core/pull/5208

usu replied:
Upgraded to api-platform 3.0.7 and fixed with https://github.com/ecamp/ecamp3/pull/3176/commits/83b49cf9435bae099aab1c57dbc625ed0d2225df

carlobeltrame commented on api/tests/Api/Activities/CreateActivityTest.php:
```diff
@@ -471,7 +470,7 @@ public function testCreateActivityValidatesMissingScheduleEntries() {
         return $this->getExamplePayload(
             Activity::class,
-            OperationType::COLLECTION,
+            '/activities',
```
Couldn't we derive the endpoint from the Entity class? I don't really see why we should have to pass this into `getExamplePayload`. From the perspective of a developer using this method, it is not clear why the payload depends on the endpoint when the Entity class and the operation are already specified in the arguments.

usu replied:
Yeah, I could probably leverage the new `OperationHelper::findOneByType` method instead of compiling the operation name manually. This however only works as long as there's a single `ApiResource` per entity class and one operation of the same type per `ApiResource`. 

Will try to refactor/simplify together with the `routePrefix` issue.

usu replied:
Simplified with https://github.com/ecamp/ecamp3/pull/3176/commits/bf6ae18430a9abe940c6ee91448f0748890e0025

carlobeltrame commented on api/tests/Api/Invitations/AcceptInvitationTest.php:
```diff
@@ -50,12 +50,13 @@ public function testAcceptInvitationSuccess() {
             ]
         );
-        $this->assertResponseStatusCodeSame(200);
+        $this->assertResponseStatusCodeSame(200); // TO DISCUSS: Wouldn't it be better to get a 204 here? The invitation doesn't really exist anymore after successful acceptance
```
204 would be fine for me

usu replied:
204 is not handled well in `hal-json-vuex`
Kept it at 200 for the moment and opened an [issue in hal-json-vuex](https://github.com/ecamp/hal-json-vuex/issues/279)

carlobeltrame commented on api/tests/Api/Invitations/AcceptInvitationTest.php:
```diff
@@ -50,12 +50,13 @@ public function testAcceptInvitationSuccess() {
+        $this->assertResponseStatusCodeSame(200); // TO DISCUSS: Wouldn't it be better to get a 204 here? The invitation doesn't really exist anymore after successful acceptance
         $this->assertJsonContains([
+            /*
             'campId' => $campCollaboration->camp->getId(),
             'campTitle' => $campCollaboration->camp->title,
             'userDisplayName' => 'Bi-Pi',
-            'userAlreadyInCamp' => false,
+            'userAlreadyInCamp' => false, */
```
Reminder: Remove or uncomment the commented payloads in this test and RejectInvitationTest

usu replied:
Cleanup done in https://github.com/ecamp/ecamp3/pull/3176/commits/2a3d2ff4c1732d72fe0f52961765880892dd2da4#diff-74c41eaa1f7f42a90ba91740dcde7f1ca280e0d6960a0061070c863e0bf4a4be

carlobeltrame commented on api/tests/Api/Invitations/RejectInvitationTest.php:
```diff
@@ -33,12 +33,13 @@ public function testRejectInvitationSucceedsWhenNotLoggedIn() {
             ]
         );
-        $this->assertResponseStatusCodeSame(200);
+        $this->assertResponseStatusCodeSame(200); // // TO DISCUSS: Wouldn't it be better to get a 204 here? The invitation doesn't really exist anymore after successful acceptance
```
Yes, as mentioned above, I would say even a DELETE request would be appropriate for this operation

usu replied:
Cleaned up in https://github.com/ecamp/ecamp3/pull/3176/commits/2a3d2ff4c1732d72fe0f52961765880892dd2da4#diff-74c41eaa1f7f42a90ba91740dcde7f1ca280e0d6960a0061070c863e0bf4a4be

carlobeltrame commented on api/tests/Serializer/PreventAutomaticEmbeddingPropertyMetadataFactoryTest.php:
```diff
@@ -16,22 +17,31 @@ class PreventAutomaticEmbeddingPropertyMetadataFactoryTest extends TestCase {
         $decorated = $this->createMock(PropertyMetadataFactoryInterface::class);
-        $propertyMetadata = new PropertyMetadata(
-            $this->createMock(Type::class),
+        $propertyMetadata = new ApiProperty(
```
Rename variable

usu replied:
Fixed with https://github.com/ecamp/ecamp3/pull/3176/commits/5e83943cfc2001c520ea438cfea130774ac2831e

carlobeltrame reviewed (commented):
We should also update https://github.com/ecamp/ecamp3/wiki/architecture-backend after merging this.

carlobeltrame committed:
Can't mock final class, just instantiate it

usu committed:
Merge pull request #13 from carlobeltrame/chore/api-platform-3.0

Use more mocking instead of allowing to override final classes

usu committed:
upgrade to api-platform 3.0.4

usu committed:
small documentation changes as per PR review

usu committed:
rename propertyMetadata to apiProperty

usu committed:
upgrade to api-platform 3.0.5

BacLuc commented on api/config/services.yaml:
```diff
@@ -167,3 +174,27 @@ services:
+    # override API-platform memory cache configuration
+    # https://github.com/api-platform/core/issues/4975#issuecomment-1253617780
+    api_platform.cache.metadata.property:
+        parent: cache.system
```
Does this enable caching on the system (e.g. container) level?
If so we have to look what this means if we have multiple pods running

usu replied:
This is the metadata cache, there's no need need to share this across pods. This config can probably be removed anyway with the next patch version. See here for more details: https://github.com/api-platform/core/issues/4975#issuecomment-1331935076

I think what you mean is HTTP cache as described here: https://api-platform.com/docs/core/performance/
This would of course be a completely different animal and needs sharing across pods.

BacLuc commented on api/src/Doctrine/Filter/ExpressionDateTimeFilter.php:
```diff
@@ -169,7 +178,7 @@ protected function compileExpression(
             // Replace each instance of {xyz.abc} with its respective joined alias
             foreach ($relations as $relation) {
-                [$joinAlias, $property] = $this->addJoinsForNestedProperty($relation, $alias, $queryBuilder, $queryNameGenerator);
+                [$joinAlias, $property] = $this->addJoinsForNestedProperty($relation, $alias, $queryBuilder, $queryNameGenerator, $resourceClass, Join::INNER_JOIN);
```
do we really want an inner join here?

BacLuc commented on api/src/DTO/ResetPassword.php:
```diff
@@ -2,38 +2,43 @@
-        ],
-    ],
-    routePrefix: '/auth/reset_password'
+    routePrefix: '/auth'
```
why didnt you keep the common url subpath /reset_password here?

usu replied:
Originally because of https://github.com/api-platform/core/issues/5203

But I would still keep it as it is now even after https://github.com/api-platform/core/issues/5203 is fixed. This is more consistent with the auto-generated operations, where `uriTemplate` starts with the resource name and `routePrefix` is only used for additional namespacing.

BacLuc commented on api/src/State/ContentNode/ContentNodePersistProcessor.php:
```diff
@@ -0,0 +1,29 @@
+    public function onBefore($data, Operation $operation, array $uriVariables = [], array $context = []): ContentNode {
+        $data = parent::onBefore($data, $operation, $uriVariables, $context);
+
+        if ($operation instanceof Post) {
```
Couldn't we just register this processor only on the Post Operation?

usu replied:
In that case, something like the `SingleTextPersistProcessor` is not possible. Which is nice, if I have exactly the same (or a very similar) logic for both create and update.

BacLuc commented on api/src/State/ActivityCreateProcessor.php:
```diff
@@ -1,30 +1,28 @@
 use Doctrine\ORM\EntityManagerInterface;
 
-class ActivityDataPersister extends AbstractDataPersister {
+final class ActivityCreateProcessor extends AbstractPersistProcessor {
```
why is this processor final?

usu replied:
No specific reason. Can remove it. Or should I mark all processors/providers as final?

BacLuc commented on api/tests/Serializer/Normalizer/RelatedCollectionLinkNormalizerTest.php:
```diff
@@ -265,23 +261,6 @@ public function testNormalizeDoesntReplaceWhenFilterDoesntApplyToMappedProperty(
         $this->shouldNotReplaceChildren($result);
     }
 
-    public function testNormalizeDoesntReplaceWhenStrategyIsNotExact() {
```
what happened here?

carlobeltrame replied:
See https://github.com/usu/ecamp3/pull/13
This test has always been flawed, because when filtering by a relation, the strategy of the search filter is always "exact". Before this test used to pass because of other reasons, and when I changed the mocking, it stopped working. So I removed the test.

BacLuc commented on api/tests/Serializer/PreventAutomaticEmbeddingPropertyMetadataFactoryTest.php:
```diff
@@ -2,12 +2,13 @@
-use ApiPlatform\Core\Metadata\Property\Factory\PropertyMetadataFactoryInterface;
-use ApiPlatform\Core\Metadata\Property\PropertyMetadata;
-use ApiPlatform\Core\Metadata\Property\SubresourceMetadata;
+// use ApiPlatform\Metadata\Property\Factory\PropertyMetadataFactoryInterface;
+// use ApiPlatform\Metadata\Property\PropertyMetadata;
+// use ApiPlatform\Metadata\Property\SubresourceMetadata;
```
i would remvoe these lines

usu replied:
Fixed in https://github.com/ecamp/ecamp3/pull/3176/commits/7f42c2c199666c77d5f0c564570ab7aeef04d28e

BacLuc commented on api/src/Metadata/Resource/OperationHelper.php:
```diff
@@ -0,0 +1,35 @@
+class OperationHelper {
+    public static function findOneByType(ResourceMetadataCollection $collection, string $className): ?Operation {
+        $it = $collection->getIterator();
+        $metadata = null;
```
You can remove this unused variable

usu replied:
Fixed in https://github.com/ecamp/ecamp3/pull/3176/commits/7f42c2c199666c77d5f0c564570ab7aeef04d28e

BacLuc commented on api/src/Metadata/Resource/OperationHelper.php:
```diff
@@ -0,0 +1,35 @@
+use ApiPlatform\Metadata\Resource\ResourceMetadataCollection;
+
+class OperationHelper {
+    public static function findOneByType(ResourceMetadataCollection $collection, string $className): ?Operation {
```
I would call the paramter $className $operationClassName

usu replied:
Fixed in https://github.com/ecamp/ecamp3/pull/3176/commits/7f42c2c199666c77d5f0c564570ab7aeef04d28e

usu committed:
additional test coverage for RelatedCollectionLinkNormalizer

usu committed:
add unit test for OperationHelper

usu committed:
minor cleanups/renamings

usu committed:
merge devel

usu committed:
test for ActivityCreateProcessor

usu committed:
CampCollaboration: remove logic from MailSerice + write tests for state processors

usu committed:
Camp: tests for state processors

usu committed:
eliminate deprecation + phpstan/pslam warnings

usu committed:
Category: state processor tests

usu committed:
Period: tests for state processor

usu committed:
User: tests for state processors

usu committed:
Profile: tests for state processors

usu committed:
test for ContentNode state processors

usu committed:
tests for DTO state processors

usu committed:
tests for abstract state processors + cleanup

usu committed:
increase test coverage for state processors

usu committed:
merge devel

usu committed:
upgrade api-platform/core to 3.0.7

usu committed:
revert routePrefix/endpoint splitup

usu committed:
simplify example payload generation

usu committed:
remove metadata cache config (resolved with api-platform/core 3.0.7)

usu committed:
remove Get operation from generic /content_nodes endpoint

usu committed:
fix root route + fix remaining tests

usu committed:
fix tests for UriTemplateNormalizer + add new failing test for ContentNodes self link inconsistency

usu committed:
upgrade api-platform to 3.0.8

usu committed:
add back metadata cache config

usu committed:
remove usage of ApiProperty:default

usu commented:
@ecampcore I think this would be ready for another check-up from you guys. Major comments should all be implemented. Last pending point is sanity check on performance. On my local machine I got inconsistent result in both directions. Want to check on the deployment directly, but not really working at the moment. Will try again later.

BacLuc commented on api/tests/State/InvitationAcceptProcessorTest.php:
```diff
@@ -0,0 +1,118 @@
+        self::assertThat($this->campCollaboration->inviteEmail, self::isNull());
+    }
+
+    // public function testUpdatesInvitationCorrectlyOnReject() {
```
could you delete that before merging?

BacLuc reviewed (approved):
Looks good to me.
Just found the one commented test which was moved, so you could delete it now.

Looked at the coverage, and its still here.
(Don't know where the -0.3 percent are lost).

Could not go through anything very thorough, but i think with another approve it will be good enough.

Thank you very much @usu

carlobeltrame reviewed (approved):
Thanks a lot!

usu committed:
merge devel

usu committed:
remove dead test code

usu committed:
update composer.lock

manuelmeister commented:
:partying_face:
